signac/contrib/job.py

```diff
@@ -359,11 +355,36 @@ def _statepoint_filename(self):
         # use str.join with os.sep instead of os.path.join for speed.
         return os.sep.join((self.workspace(), self.FN_MANIFEST))
 
-    @property
+    # Decorated properties aren't supported: https://github.com/python/mypy/issues/1362
+    @property  # type: ignore
+    @deprecated(
+        deprecated_in="1.8",
+        removed_in="2.0",
+        current_version=__version__,
+        details="Use Job.path instead.",
+    )
     def ws(self):
-        """Alias for :meth:`~Job.workspace`."""
-        return self.workspace()
+        """Alias for :meth:`~Job.path`."""
+        return self.path
 
+    @property
+    def path(self):
+        """str: The path to the job directory.
+
+        See :ref:`signac job -w <signac-cli-job>` for the command line equivalent.
+        """
+        if self._wd is None:
```

Comment:
I'd like to rename `self._wd` to `self._path` to match.

Reply:
Yeah we should do that, but you may have noticed that I made a separate note elsewhere in this PR that we should explore using `functools.cached_property` instead of manually maintaining these variables. I'd like to try that first since that would remove these variables altogether, if you and other reviewers agree.

Reply:
And require Python 3.8, when that feature was introduced?

Reply:
I think we have to start requiring Python 3.8 anyway as of #658, right? We just need to update the changelog/setup.py/etc to make that official. But yes, that would require more changes. And to be clear, I'm proposing to do that in a separate PR.

Reply:
We support older numpy versions that are compatible with Python < 3.8. We just can’t test “latest numpy” with all the Python versions we currently support.

Reply:
Python 3.7 support is dropped according to NEP 29.

Reply:
Yeah I would think #658 is a reason to require Python 3.8 because that PR updated a dependency (Numpy) to a version that requires 3.8 directly as a consequence of NEP 29. [We document that we follow NEP 29](https://docs.signac.io/projects/core/en/latest/installation.html?highlight=NEP#installation), and I think every time this debate has come up we've agreed that we should follow it. By that we really should have dropped Python 3.6 almost 2 years ago, but our docs still say that they support it. I think starting with signac 2.0 we can make the move to strictly following NEP 29.

Reply:
That said, I'll go ahead and rename these variables for now.

Reply:
> I would think https://github.com/glotzerlab/signac/pull/658 is a reason to require Python 3.8 because that PR updated a dependency (Numpy) to a version that requires 3.8 directly as a consequence of NEP 29.

If the next version of some dependency required Python 3.10, we wouldn’t need to require Python 3.10 because we can always use an older version of that dependency for testing with Python versions <3.10. The same applies here — I don’t think #658 is a valid justification to require Python 3.8. The justification is that we want to use Python 3.8 features (and that’s the highest we can go while being NEP 29 compliant).

Reply:
Put differently, “strictly following NEP 29” just means we drop support _no faster_ than it allows. It’s an “infimum,” in the sense of a greatest lower bound.

In some cases, it has been easier to leave our CI alone and not increase our requirements immediately when NEP 29 drops support for something.

Reply:
I disagree that strictly following NEP 29 means treating it as a "greatest lower bound". [Having different packages decide to drop support for Python/NumPy versions at different times is explicitly discouraged by the NEP](https://numpy.org/neps/nep-0029-deprecation_policy.html#ad-hoc-version-support). A key goal is to create consistent expectations of what language/library features are available for use across the ecosystem. In practice, most packages (including NEP authors like Matplotlib, scikit-learn, and IPython) are treating the NEP's timeline as an infimum, so I'm OK with doing the same. However, that comes at the cost of creating more inconsistent expectations across the Python ecosystem, which is precisely what the NEP is designed to solve, and we lose clarity on what constitutes a "good enough reason" for a version bump.

If we are OK with that, then the question is how we make that decision. Do we wait until the NEP 29 window has passed and then bump whenever we see a new language/NumPy feature that we think is worthwhile? Do we standardize in any way, or do we make this decision on an ad hoc basis from version to version? Are there other constraints (e.g. the ability to use the Python versions available on supercomputers) that provide a more conservative guide than NEP 29's relatively aggressive policy?

Reply:
These statements from NEP 29 are a little conflicting (emphasis added):
> When a project releases a new major or minor version, we recommend that they support _at least_ all minor versions of Python introduced and released […]

> Minimum Python and NumPy version support should be _adjusted upward on every major and minor release_, but never on a patch release.

I was aware of the first statement’s “at least” but was not aware of the second statement’s recommendation to (frequently!) bump Python/NumPy version minimum requirements. That explains part of why we had different interpretations of this piece of NEP 29. 😅

> Do we wait until the NEP 29 window has passed and then bump whenever we see a new language/NumPy feature that we think is worthwhile?

This is my recommendation. Bumping on every major/minor release would be a pain. Minimizing maintenance costs while supporting NEP 29 is my goal.

> Are there other constraints (e.g. the ability to use the Python versions available on supercomputers) that provide a more conservative guide than NEP 29's relatively aggressive policy?

This is not a high priority to me unless we have users willing to help identify and maintain such support. Most users have the ability to use a conda environment, container, or similar means to get beyond the provided system Python version. I don’t want to put that burden entirely on signac maintainers — HPC admins and users can shoulder some of it.

Reply:
Right, so basically I am fine with this policy but we just need to be careful about what we state. I don't think that this constitutes "following" NEP 29, but I think it's consistent with how most packages are treating the proposal. I think our main action item is to amend the statement [here](https://docs.signac.io/projects/core/en/latest/installation.html?highlight=nep%2029#installation), [here](https://docs.signac.io/projects/dashboard/en/latest/installation.html?highlight=nep%2029#installation) and [here](https://docs.signac.io/projects/flow/en/latest/installation.html?highlight=nep%2029#installation) to say something like "signac uses NEP 29 as a guideline for when to drop support for Python and NumPy versions and does not guarantee support beyond the versions recommended in that proposal". That is in line with [Scipy's statement](https://docs.scipy.org/doc/scipy-1.8.0/html-scipyorg/dev/toolchain.html#python-versions), which to my knowledge is the only such statement made by any of the major Python packages behind NEP 29.

Reply:
I addressed this in https://github.com/glotzerlab/signac/pull/715/commits/1b18ffab3f3a5497b7911f4e82f089ca4242e989.